numpy/lib/function_base.py

```diff
@@ -84,15 +84,34 @@ def _hist_optim_numbins_estimator(a, estimator):
     A description of the estimators can be found at
     https://en.wikipedia.org/wiki/Histogram#Number_of_bins_and_width
 
+    Parameters
+    ----------
+    a : array_like
+        The data with which to estimate the number of bins
     estimator: str
         If ``estimator`` is one of ['auto', 'fd', 'scott', 'doane',
         'rice', 'sturges', 'sqrt'], this function will choose the
         appropriate estimation method and return the optimal number of
         bins it calculates.
+    data_range: tuple (min, max)
+        The range that the data to be binned should be restricted to.
+    data_weights:
+        weights are not supported, so this field must be empty or None.
     """
     if a.size == 0:
         return 1
 
+    if data_weights is not None:
+        raise TypeError("Automated estimation of the number of "
+                        "bins is not supported for weighted data")
+
+    if data_range is not None:
+        mn, mx = data_range
+        keep = (a >= mn)
+        keep &= (a <= mx)
+        if not np.logical_and.reduce(keep):
```

Comment:
Previous discussion with @nayyarv's comment was here: https://github.com/numpy/numpy/pull/6288#discussion_r38879392

I.e., this file uses `and.reduce` elsewhere for this same computation. I'm not too bothered either way, using `add.reduce` consistently isn't really worse than using `all` in some places and `add.reduce` in others, and I'd like to get this moving for 1.11, so @shoyer -- I'm going to merge this and we can figure out whether we care about `add.reduce` versus `all` as a separate matter? Feel free to squawk if you think this is a grave mistake :-)